### ISSUE TITLE
has_many_as.build does not set the inverse polymorphic belongs_to assoc on the new model

### ISSUE THREAD

**Issue by bughit** [rails/rails#21843]
as a result of that the `required: true` validation on the child fails when the parent is saved (comment out the `assert comment.commentable` line in the test)

and in general it's very misleading that it appears disconnected, there should be enough info to connect it.

``` ruby
begin
  require 'bundler/inline'
rescue LoadError => e
  $stderr.puts 'Bundler version 1.10 or later is required. Please update your Bundler'
  raise e
end

gemfile(true) do
  source 'https://rubygems.org'
  # Activate the gem you are reporting the issue against.
  gem 'activerecord', '4.2.4'
  gem 'sqlite3'
end

require 'active_record'
require 'minitest/autorun'
require 'logger'

# Ensure backward compatibility with Minitest 4
Minitest::Test = MiniTest::Unit::TestCase unless defined?(Minitest::Test)

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: 'sqlite3', database: ':memory:')
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
  end

  create_table :comments, force: true do |t|
    t.belongs_to :commentable, polymorphic: true
  end
end

class Post < ActiveRecord::Base
  has_many :comments, as: :commentable
end

class Comment < ActiveRecord::Base
  belongs_to :commentable, polymorphic: true, required: true
end

class BugTest < Minitest::Test
  def test_association_stuff

    post = Post.new
    comment = post.comments.build
    assert comment.commentable
    post.save!

  end
end
```

**Comment by bughit** [comment]
same for `post.comments << Comment.new`

**Comment by maclover7** [comment]
Reproduction script fails on master.

**Comment by fschwahn** [comment]
@maclover7 I've run into this issue as well. The following reproduction script works with rails 5.0.1 as well as master - the test fails for both.

Note that the owner is correctly set when the post was already persisted, it only fails to attach the models when the parent is a new record (`Comment` in this example).

I've included a non-polymorphic association as well to demonstrate that it works there as expected. It's not actually relevant to the test.

```ruby
#!/usr/bin/env ruby

require "minitest/autorun"

begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"
  gem "activerecord", "5.0.1", require: "active_record"

  # gem "activerecord", github: "rails/rails", require: "active_record"
  # gem "arel", github: "rails/arel"

  gem "sqlite3"
end

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :posts, force: true do |t|
    t.string :title
  end

  create_table :comments, force: true do |t|
    t.integer :post_id
    t.text :body
  end

  create_table :addresses, force: true do |t|
    t.integer :owner_id
    t.string :owner_type
    t.text :place
  end
end

class Post < ActiveRecord::Base
  has_many :comments
  has_many :addresses, as: :owner
end

class Comment < ActiveRecord::Base
  belongs_to :post
end

class Address < ActiveRecord::Base
  belongs_to :owner, polymorphic: true
end

class BugTest < Minitest::Test
  def test_polymorphic_build_on_new_record
    post = Post.new
    comment = post.comments.build(body: "foo")
    assert_equal comment.post, post

    address = post.addresses.build(place: "time square")
    assert_equal address.owner, post
  end
end
```

**Comment by maclover7** [comment]
Investigating.

**Comment by kamipo** [comment]
Fixed by #28808.
